### PR TITLE
irmin-bench: Factorised replayable trace

### PR DESCRIPTION
**Pull request by Ngoguey42**
on top of #1357, extracted from #1332

### CODE REVIEW

**Comment by Ngoguey42** [comment]
@zshipko @gbelouze I'm going to open several PRs that upstream bits of code that you could find interest in reusing for your own "trace" use cases.

In this PR you will find in [`trace_common.ml`](https://github.com/mirage/irmin/pull/1358/files#diff-294dc57a9868feb44aafcc636e117c1b755a2f90e99726e8adcc945eb7427bfeR1) a way to simplify the creation of a trace file format, [`Trace_definitions.Replayable_trace`](https://github.com/mirage/irmin/pull/1358/files#diff-e8ae98eb8dc81df6b5a8eb573d417eb1a692f2d61b3af715309754c2c33bc67fR47-R117
) being an example of such a file format.

Feel free to review the whole PR if you have time :)

**Review by zshipko — APPROVED** [review]
This looks good - duplicating `tree.ml` was something I've been concerned about, so thank you! 

I don't have anything else to add yet but perhaps something will come up as I continue to look at it.

File: bench/irmin-pack/trace_common.ml [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    - Magic (Magic.t, 8 bytes)
+    - Version (int32, 4 bytes)
+    - Length of header (varint, >=1 byte)
+    - Header (header_t, _ bytes)
+    - Arbitrary long series of rows, of unspecified length, each prefixed by their length:
+      - Length of row (varint, >=1 byte)
+      - Row (row_t, _ bytes)
+    v}
+
+    This file is also meant to be used from Tezos. OCaml version 4.09 and the
```
Comment (icristescu) [inline bench/irmin-pack/trace_common.ml:14]:
```suggestion
    This file is also meant to be used from Tezos. OCaml version 4.11.0 and the
```

File: bench/irmin-pack/trace_common.ml [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+(** Contains everything needed to read a file as if it is written with the
+    lastest version. *)
+
+(** A box containing the above record *)
+type ('latest_header, 'latest_row) version_converter =
+  | Version_converter :
+      ('latest_header, 'latest_row, 'header, 'row) version_converter'
+      -> ('latest_header, 'latest_row) version_converter
+
+module type FILE_FORMAT = sig
```
Comment (icristescu) [inline bench/irmin-pack/trace_common.ml:125]: change the all caps?

File: bench/irmin-pack/trace_definitions.ml [hunk 3]
```diff
@@ -0,0 +1,118 @@
+(** Traces file format definitions.
+
+    This file is also meant to be used from Tezos. OCaml version 4.09 and the
+    32bit architecture should be supported.
+
+    {3 Traces Workflow}
+
+    A Tezos node (may) output a [Replayable_trace] file (TODO: Or a Raw traces,
```
Comment (icristescu) [inline bench/irmin-pack/trace_definitions.ml:8]: a replayable trace is a raw trace + postprocessing? maybe reformulate to say that

**Review by icristescu — APPROVED** [review]
Thanks, LGTM.

File: bench/irmin-pack/trace_common.ml (hunk 1 above)
Comment (icristescu) [reply bench/irmin-pack/trace_common.ml:14]: my mistake, ignore this :)

**Pull request merged**